### PR TITLE
Recipes quality search

### PR DESCRIPTION
**Pull request by markus1189**
I am still fairly new to the game (and the code base), but I found myself wondering why I can't search for recipes that will create a tool that has a specific quality, e.g. butchering.

So I expanded the possible search prefixes for recipe search to add `q:` for searching qualities.
Quality search in action

I separated this into 2 commits because the last one (15c5624) also changes the description as seen in the screenshot above, however it will lack the translations which might be a blocking issue?

Best,
Markus

### CODE REVIEW

File: src/crafting.cpp [hunk 1]
```diff
@@ -863,8 +863,19 @@ const recipe *select_crafting_recipe( int &batch_size )
             redraw = true;
             keepline = true;
         } else if (action == "FILTER") {
+          std::string desc = std::string("Special prefixes:\n") +
```
Comment (BevapDin) [inline src/crafting.cpp]: C++ combines adjacent C-String literals into one C-String literal:

```C++
std::string text = "AB" "CD" "EF"
"GH"
"IJ";
// text contains now "ABCDEFGHIJ"
```

For translation: add `_(` and `)` around the C-String literal:

```C++
std::string text = _("this will be translated");
```

The `_` marks the literal as to be translated (there is a script that extracts all marked strings, they can be translated separately). It also calls the function to get the translation at runtime.

However, this `_(desc.c_str())` won't work because there is no string to extract here.

File: src/crafting.cpp [hunk 2]
```diff
@@ -1293,6 +1307,20 @@ void pick_recipes(const inventory &crafting_inv,
                         continue;
                     }
                 }
+                if(search_qualities) {
+                  bool match_found = false;
+                  itype *it = item::find_type(rec->result);
+                  for(std::map<std::string, int>::const_iterator quality = it->qualities.begin(); quality != it->qualities.end() && !match_found; ++quality) {
+                    std::string qname = quality::get_name(quality->first).c_str();
```
Comment (BevapDin) [inline src/crafting.cpp]: `get_name` returns a `std::string`, no need for the `.c_str()`.

**Comment by Lain-** [comment]
Thanks! This will be useful.

**Comment by markus1189** [comment]
Thanks for the feedback!  What remains to be done for the translation part, i.e. how would I add the german translation?  I found the `.po` files, but I am note sure if some of that is auto generated (e.g. `#: src/crafting.cpp` makes me suspicious :)

**Comment by BevapDin** [comment]
> What remains to be done for the translation part, i.e. how would I add the german translation? I found the  .po  files, but I am note sure if some of that is auto generated (e.g.  #: src/crafting.cpp  makes me suspicious :)

You don't edit those directly. They are updated through a script (if you care, the stuff is located in the lang folder, in this case the `lang/update_pot.sh` is used). However, (as far as I know) that is done automatically, by an external site (Transifex) and the awesome translators use that site to update the translations, which is than periodically merged into this repository.

tl;dr: The programmer does not update the translation data.

File: src/crafting.cpp [hunk 3]
```diff
@@ -1293,6 +1307,20 @@ void pick_recipes(const inventory &crafting_inv,
                         continue;
                     }
                 }
+                if(search_qualities) {
+                  bool match_found = false;
+                  itype *it = item::find_type(rec->result);
+                  for(std::map<std::string, int>::const_iterator quality = it->qualities.begin(); quality != it->qualities.end() && !match_found; ++quality) {
```
Comment (BevapDin) [inline src/crafting.cpp]: Instead of writing the iterator type yourself, you could use `auto` and let the compiler figure it out.

It could also be done with a (equivalent) range-based loop:
```C++
for( auto & e : it->qualities ) {
   quality::get_name( e.first );
   ...
```

Comment (narc0tiq) [inline src/crafting.cpp]: ^ Equivalent range-based loop plus `if(match_found) { break; }`.

Also, iterators are why `auto` was invented, I'm pretty sure (only slightly tongue-in-cheek).

File: src/crafting.cpp [hunk 4]
```diff
@@ -1293,6 +1307,20 @@ void pick_recipes(const inventory &crafting_inv,
...
+                    std::string qname = quality::get_name(quality->first);
+                    if (lcmatch(qname, filter)) {
+                      match_found = true;
+                    }
+                  }
+
+                  if (!match_found) {
```
Comment (BevapDin) [inline src/crafting.cpp]: The check is somehow pointless, the loop would continue anyway (until it's stopped by the loop condition), it's like `if(...) continue; else continue;`.

**Comment by markus1189** [comment]
That range based loop is much better you're right, I normally don't write any C++ so any suggestions are welcome.  I first thought I could use `lcmatch_any` to avoid the use of the iterator and boolean var completely, but unfortunately that is only for `vector<vector<T>>`.

File: src/crafting.cpp [hunk 5]
```diff
@@ -863,8 +863,19 @@ const recipe *select_crafting_recipe( int &batch_size )
...
+            "  [s] search skills\n" +
+            "  [S] search skill used only\n" +
+            "Examples:\n" +
+            "  t:hammer\n" +
+            "  c:two by four\n" +
+            "  q:butchering\n" +
+            "  s:cooking";
             filterstring = string_input_popup(_("Search:"), 85, filterstring,
-                                              _("Search tools or component using prefix t. \nSearch skills using prefix s, or S for skill used only. \n (i.e. \"t:hammer\" or \"c:two by four\" or \"s:cooking\".)"));
+                                              _(desc.c_str()));
```
Comment (KA101) [inline src/crafting.cpp]: OK, this'll need reworked before we can merge, see BevapDin's on translation protocol.

**Comment by markus1189** [comment]
Whoops I missed that when squashing the original two commits into one, fixed. :)

**Comment by KA101** [comment]
Ooops, no, looks like a bad git update there.  Someone more conversant with git cherry-pick might be able to grab something useful from this; otherwise, gonna need to close, fresh branch from master, and re-implement/re-PR.  Sorry, markus1189.  :-(

**Comment by markus1189** [comment]
I might be able to help, what's the problem?  I rebased the commit onto 9db58e1 of master without conflicts

**Comment by markus1189** [comment]
Okay there you go, I did a cherry pick from master(df6647a) and that worked.  :)

**Comment by KA101** [comment]
If you look at the PR on the Gite, sure, no conflicts, but each and every commit in the interval you updated is now duplicated in the PR.  Not the first time that's happened, likely not the last either.  If I were to merge this PR as it stands, it'd louse up our repo history but good.

**Comment by markus1189** [comment]
Which should no longer be the case (if you have not seen my comment above) :)

**Comment by KA101** [comment]
It was when I posted; now, correct, it is not.  Yay updating Git site.
